### ISSUE TITLE
[idea] implement auto-diff example

### ISSUE THREAD

**Issue by hovsep** [hovsep/fmesh-examples#11]
// Model holds the parameter we want to learn: weight W
type Model struct {
	W float64
}

func main() {
	// Seed random generator for varied training data
	rand.Seed(time.Now().UnixNano())

	// Initialize model with a guess for W
	model := &Model{W: 0.5}

	// Learning rate controls how big each update step is
	lr := 0.01

	// Number of training iterations
	steps := 5000

	// Forward component:
	// - Receives input pair (x, y)
	// - Predicts yPred = W * x using current model.W
	// - Calculates squared loss = (y - yPred)^2
	// - Emits (x, y, yPred, loss) to next component
	forward := component.New("Forward").
		WithInputs("in").
		WithOutputs("out").
		WithActivationFunc(func(c *component.Component) error {
			in := c.InputByName("in").FirstSignalPayloadOrDefault(nil)
			if in == nil {
				return nil // no input, do nothing
			}
			data := in.([2]float64)
			x, y := data[0], data[1]

			yPred := model.W * x
			loss := (y - yPred) * (y - yPred)

			// Send all values forward as one payload
			c.OutputByName("out").PutSignals(signal.New([4]float64{x, y, yPred, loss}))
			return nil
		})

	// Backward component:
	// - Receives (x, y, yPred, loss)
	// - Calculates gradient of loss w.r.t W: grad = -2 * x * (y - yPred)
	// - Emits gradient and loss forward for updater
	backward := component.New("Backward").
		WithInputs("in").
		WithOutputs("out").
		WithActivationFunc(func(c *component.Component) error {
			in := c.InputByName("in").FirstSignalPayloadOrDefault(nil)
			if in == nil {
				return nil
			}
			data := in.([4]float64)
			x, y, yPred, loss := data[0], data[1], data[2], data[3]

			grad := -2 * x * (y - yPred)

			c.OutputByName("out").PutSignals(signal.New([2]float64{grad, loss}))
			return nil
		})

	// Updater component:
	// - Receives (grad, loss)
	// - Updates model.W using gradient descent: W -= lr * grad
	// - Prints current loss and parameter value
	updater := component.New("Updater").
		WithInputs("in").
		WithActivationFunc(func(c *component.Component) error {
			in := c.InputByName("in").FirstSignalPayloadOrDefault(nil)
			if in == nil {
				return nil
			}
			data := in.([2]float64)
			grad, loss := data[0], data[1]

			// Update the weight parameter
			model.W -= lr * grad

			fmt.Printf("Loss: %.4f, w: %.4f\n", loss, model.W)
			return nil
		})

	// Wire components: forward -> backward -> updater
	forward.Outputs().ByName("out").PipeTo(backward.Inputs().ByName("in"))
	backward.Outputs().ByName("out").PipeTo(updater.Inputs().ByName("in"))

	// Create the mesh with all components
	mesh := fmesh.New("AutoDiffDemo").
		WithComponents(forward, backward, updater)

	// Training loop: run many steps
	for i := 0; i < steps; i++ {
		// Generate random training example
		x := rand.Float64()*10 - 5 // x in [-5,5]
		y := 2 * x                 // true function: y = 2*x

		// Clear previous input signals to avoid mixing data between steps
		forward.Inputs().ByName("in").Clear()

		// Send new input (x,y) to the forward component
		forward.Inputs().ByName("in").PutSignals(signal.New([2]float64{x, y}))

		// Run the entire mesh once to process this example
		_, err := mesh.Run()
		if err != nil {
			fmt.Println("Error:", err)
			break
		}
	}

	// Print the final learned parameter
	fmt.Printf("Final learned w = %.4f (should approach 2)\n", model.W)
}
